spring-data-mongodb/src/main/java/org/springframework/data/mongodb/util/json/ParameterBindingJsonReader.java

```diff
@@ -124,10 +130,20 @@ public ParameterBindingJsonReader(String json, ParameterBindingContext bindingCo
 
 		this.bindingContext = bindingContext;
 
-		Matcher matcher = PARAMETER_ONLY_BINDING_PATTERN.matcher(json);
+		Matcher matcher = ENTIRE_QUERY_BINDING_PATTERN.matcher(json);
 		if (matcher.find()) {
-			currentValue = bindableValueFor(new JsonToken(JsonTokenType.UNQUOTED_STRING, json)).getValue();
+			BindableValue bindingResult = bindableValueFor(new JsonToken(JsonTokenType.UNQUOTED_STRING, json));
+			try {
+				if (bindingResult.getType().equals(BsonType.DOCUMENT)) {
+					currentValue = Document.parse(bindingResult.getValue().toString());
```

Comment:
This part throws an error when the bound value is an `org.bson.Document` or a `Map`. `StringBasedMongoQueryUnitTests` do not pass at this point.

Reply:
Right, so make more sense to check if expression result is a String